### PR TITLE
Add castling with the king

### PR DESCRIPTION
**Pull request by abdullahalam**
To support castling with the king, and not sure about the testing!

### CODE REVIEW

**Comment by abdullahalam** [comment]
@jacobwgillespie To support castling with the king 

- To need one method that checks to see if a king is allowed to castle in a given direction 
- After created with castling with king method, the king hasn't moved, the rook hasn't moved, and there are no pieces in between 
- A method that actually executes the castling, updating the position of both the king and the rook.

File: app/models/king.rb [hunk 1]
```diff
@@ -12,6 +12,39 @@ def valid_move?(x, y)
...
+  def castling_kingside?(x, y)
+    x == 6 && y == y_position && castling_rook_kingside && castling_rook_kingside.castling_never_moved? && \
+      castling_rook_kingside.blocked?(4, y_position) == false
+  end
+
+  # The rook to castle with king side
+  def castling_rook_kingside
+    can_castling_rook_kingside = game.pieces.where(x_position: 0, y_position: y_position, \
+                                                   color: 'white' || 'black', type: 'Rook')
+    @castling_rook_kingside = can_castling_rook_kingside.first
```
Comment (jacobwgillespie) [inline app/models/king.rb]: I don't think this needs to be saved to an instance variable, so this line could instead just be:

```ruby
can_castling_rook_kingside.first
```

However then the entire method could just be the query itself:

```ruby
def castling_rook_kingside
  game.pieces.where(x_position: 0, y_position: y_position, color: color, type: 'Rook').first
end
```

File: app/models/king.rb [hunk 2]
```diff
@@ -12,6 +12,39 @@ def valid_move?(x, y)
...
+
+  # The king can attack to the castle on the king side
+  def castling_kingside?(x, y)
+    x == 6 && y == y_position && castling_rook_kingside && castling_rook_kingside.castling_never_moved? && \
+      castling_rook_kingside.blocked?(4, y_position) == false
+  end
+
+  # The rook to castle with king side
+  def castling_rook_kingside
+    can_castling_rook_kingside = game.pieces.where(x_position: 0, y_position: y_position, \
```
Comment (jacobwgillespie) [inline app/models/king.rb]: I think you want to grab the rook that has the same color as the current king, so you can change `color: 'white' || 'black'` to just `color: color`.

File: app/models/king.rb [hunk 3]
```diff
@@ -12,6 +12,39 @@ def valid_move?(x, y)
...
+
+    # The king can castle
+    return true if move_castle?(x, y)
+
     true
   end
+
+  # Support castling with the king
+  def move_castle?(x, y)
+    castling_kingside?(x, y) && castling_never_moved?
```
Comment (jacobwgillespie) [inline app/models/king.rb]: This looks good, but do we want to add the `castling_queenside?` as well?

**Comment by abdullahalam** [comment]
@jacobwgillespie Here we go! Please check it again. 
- To add `castling_queenside?` in `move_castle?(x, y)` method
- To add `castling_queenside?` method (The king can attack to the castle on the queen side)
- To add `castling_rook_queenside` method (The rook to castle with queen side)
- To add if-else statement on the `castling_queenside?(x, y)` in the `castle!(x, y)` method

**Pull request merged**